tmva/tmva/inc/TMVA/BatchGenerator.hxx

```diff
@@ -0,0 +1,263 @@
+        ROOT::RDataFrame x_rdf(x_spec);
+
+        size_t progressed_events, passed_events;
+
+        // add filters if given
+        if (filters.size() > 0) {
+            auto x_filter = x_rdf.Filter(filters[0], "F1");
+
+            for (auto i = 1; i < filters.size(); i++) {
+                auto name = "F" + std::to_string(i);
+                x_filter = x_filter.Filter(filters[i], name);
+            }
+
+            // add range
+            auto x_ranged = x_filter.Range(chunk_size);
+            auto myReport = x_ranged.Report();
+
+            // load data
+            x_ranged.Foreach(func, cols);
+
+            // get the loading info
+            progressed_events = myReport.begin()->GetAll();
+            passed_events = (myReport.end()-1)->GetPass();
+        }
+        
+        // no filters given
+        else {
+            // add range
+            auto x_ranged = x_rdf.Range(chunk_size);
+            auto myCount = x_ranged.Count();
+
+            // load data
+            x_ranged.Foreach(func, cols);
+
+            // get loading info
+            progressed_events = myCount.GetValue();
+            passed_events = myCount.GetValue();
+        }
+        
+        // std::cout << "BatchGenerator::init => tensor: " << x_tensor << std::endl;
+
+        current_row += progressed_events;
+
+        // Create batches for the current_chunk.
+        // First get the correct idices to use, then turn them into batches
+        // Validation batches only have to be made in the first epoch
+        if (training_idxs.size() > current_chunk) {
+            batch_loader->CreateTrainingBatches(x_tensor, training_idxs[current_chunk]);
+        }
+        else {
+            createIdxs(current_chunk, progressed_events);
+            batch_loader->CreateTrainingBatches(x_tensor, training_idxs[current_chunk]);
+            batch_loader->CreateValidationBatches(x_tensor, validation_idxs[current_chunk]);
+        }
+    }
+    void createIdxs(size_t current_chunk, size_t progressed_events)
+    {
+        std::vector<size_t> row_order = std::vector<size_t>(progressed_events);
+
+        std::iota(row_order.begin(), row_order.end(), 0);
+
+        std::shuffle(row_order.begin(), row_order.end(),rng);
+
+        size_t num_validation = progressed_events * validation_split;
+
+        std::vector<size_t> valid_idx({row_order.begin(), row_order.begin() + num_validation});
+        std::vector<size_t> train_idx({row_order.begin() + num_validation, row_order.end()});
+
+        training_idxs.push_back(train_idx);
+        validation_idxs.push_back(valid_idx);
+    }
+
+public:
+
+    BatchGenerator(std::string _file_name, std::string _tree_name, std::vector<std::string> _cols, 
+                   std::vector<std::string> _filters, size_t _chunk_size, size_t _batch_size, std::vector<size_t> _vec_sizes = {}, double _validation_split=0.0, 
```

Comment:
Is it intended that some of the arguments are pass-by-value; passing (const) l-value references should avoid some additional unneeded copies.